src/crypto/aead/aead_macros.rs

```diff
@@ -0,0 +1,185 @@
+                           $macbytes:expr) => (
+
+use randombytes::randombytes_into;
+#[cfg(feature = "default")]
+use rustc_serialize;
+use libc::c_ulonglong;
+use std::ptr;
+
+/// Number of bytes in `Key`.
+pub const KEYBYTES: usize = $keybytes;
+
+/// Number of bytes in a `Nonce`.
+pub const NONCEBYTES: usize = $noncebytes;
+
+/// Number of bytes in the authenticator tag of an encrypted message
+/// i.e. the number of bytes by which the ciphertext is larger than the
+/// plaintext.
+pub const MACBYTES: usize = $macbytes;
+
+new_type! {
+    /// `Key` for symmetric authenticated encryption
+    ///
+    /// When a `Key` goes out of scope its contents
+    /// will be zeroed out
+    secret Key(KEYBYTES);
+}
+
+new_type! {
+    /// `Nonce` for symmetric authenticated encryption
+    nonce Nonce(NONCEBYTES);
+}
+
+/// `gen_key()` randomly generates a secret key
+///
+/// THREAD SAFETY: `gen_key()` is thread-safe provided that you have
+/// called `sodiumoxide::init()` once before using any other function
+/// from sodiumoxide.
+pub fn gen_key() -> Key {
+    let mut key = [0; KEYBYTES];
+    randombytes_into(&mut key);
+    Key(key)
+}
+
+/// `gen_nonce()` randomly generates a nonce
+///
+/// THREAD SAFETY: `gen_key()` is thread-safe provided that you have
+/// called `sodiumoxide::init()` once before using any other function
+/// from sodiumoxide.
+pub fn gen_nonce() -> Nonce {
+    let mut nonce = [0; NONCEBYTES];
+    randombytes_into(&mut nonce);
+    Nonce(nonce)
+}
+
+/// `encrypt()` encrypts the plaintext `m` using a secret
+/// key `k`, and a public nonce `n`, and emits a combined
+/// byte stream of ciphertext, and an authentication tag
+/// authenticating both `m` and the optional additional data in `ad` .
+///
+/// The public nonce should never be re-used with the same key.
+///
+/// The encrypted length will be at most m.len() + MACBYTES.
+pub fn encrypt(m: &[u8],
+               ad: &[u8],
+               &Nonce(ref n): &Nonce,
+               &Key(ref k): &Key) -> Vec<u8> {
+
+    let mut c : Vec<u8> = Vec::with_capacity(m.len() + MACBYTES);
+    let mut clen : c_ulonglong = 0;
+    unsafe {
+        $encrypt_name(
+            c.as_mut_ptr(), &mut clen,
+            m.as_ptr(), m.len() as c_ulonglong,
+            ad.as_ptr(), ad.len() as c_ulonglong,
+            ptr::null(),
+            n,
+            k);
+        assert!(clen as usize <= c.capacity());
+        c.set_len(clen as usize);
+    }
+
+    c
+}
+
+/// `encrypt()` encrypts the plaintext `in_out` of length `m_len`
+/// using a secret key `k`, and a public nonce `n`, and overwrites
+/// `in_out` with a combined byte stream of ciphertext, and an
+/// authentication tag authenticating both the plaintext and
+/// the optional additional data in `ad` .
+///
+/// The public nonce should never be re-used with the same key.
+///
+/// The encrypted length will be at most m_len + MACBYTES, so
+/// in_out.len() should be at least m_len + MACBYTES.
```

Comment:
I'm not a fan of the API of this function. I think the best bet for the `inplace()` functions (and we should do the same for `crypto::secretbox` and `crypto::box` as well is to use the detached-functions and make the API look something like:

`encrypt_inplace(m: &mut [u8], ad: &[u8], &Nonce(ref n): &Nonce, &Key(ref k): &Key) -> Result<Tag, ()>` where `Tag` is a detached authentication tag (which we always know the size of). Of course with similar changes to `decrypt_inplace()`.

We probably also need helper functions to separate the authentication tag from an authenticated encrypted message, something like `split_tag(&'a [u8]) -> Result<(&'a Tag, &'a [u8])> to be able to easily switch between the `inplace` and the allocating versions. The name `split_tag()` isn't good though.